### ISSUE TITLE
ERR_INVALID_RESPONSE 

### ISSUE THREAD

**Issue by reinhardt1053** [iusvar/grav-plugin-snappygrav#8]
I am on mac OS X El Captain, I installed snappygrav and all the dependencies through composer. :pdf and pdfcomplete links don't work, I got an ERR_INVALID_RESPONSE page.

**Comment by iusvar** [comment]
Please check the path to the `wkhtmltopdf-i386` program in `snappygrav.yaml` files and, if different from the true path, change it accordingly.

**Comment by reinhardt1053** [comment]
Is it supposed to be an absolute path?

**Comment by iusvar** [comment]
No, it's not necessary. If you are using is not a problem. It is important to refer to `wkhtmltopdf` program. For example: `wk_path: 'vendor/h4cc/wkhtmltopdf-i386/bin/wkhtmltopdf-i386'`

I do not know OS X. Check this page for the OS required version from [wkhtmltopdf](http://wkhtmltopdf.org/downloads.html)

**Comment by reinhardt1053** [comment]
I installed it on Ubuntu 16.04 and I got the same problem. I use the RTFM skeleton https://github.com/getgrav/grav-skeleton-rtfm-site

**Comment by iusvar** [comment]
Try to comment the lines 140, 144 to 152 of the file `snappygrav.php`.
You can attach a screenshot coming out?

**Comment by reinhardt1053** [comment]
First attempt:
I installed Wkhtmltopdf with composer:
    composer require h4cc/wkhtmltopdf-i386

It didn't work.

Second attempt:

I installed Wkhtmltopdf with apt:

sudo apt install wkhtmltopdf

and I changed Wkhtmltopdf Path to /usr/bin/wkhtmltopdf in the snappygrav plugin options.
<img width="852" alt="screen shot 2016-09-21 at 11 46 09" src="https://cloud.githubusercontent.com/assets/272154/18706274/a25064cc-7ff1-11e6-8c9a-f857b36ba79c.png">


I commented out the lines 140, 144 to 152 of the file snappygrav.php and I get this error when I try to generate the pdf:

<img width="1282" alt="screen shot 2016-09-21 at 11 46 30" src="https://cloud.githubusercontent.com/assets/272154/18706249/84dbb05e-7ff1-11e6-812c-1884b12585b7.png">

**Comment by iusvar** [comment]
Have you run the `composer require knplabs/knp-snappy` command ?
Check that there is the `vendor/knplabs/knp-snappy/src/Knp/Snappy` folder.

**Comment by reinhardt1053** [comment]
Ok, I executed the commands from the grav root folder:

 ` composer require knplabs/knp-snappy `

The folder vendor/knplabs/knp-snappy/src/Knp/Snappy is there now.

I removed 

wkhtmltopdf installed by apt because apparently it doesn't work in an headless environment (I am running Ubuntu 16.04 Server)

http://unix.stackexchange.com/questions/192642/wkhtmltopdf-qxcbconnection-could-not-connect-to-display

And I installed it with composer from the grav root folder:

`composer require h4cc/wkhtmltopdf-i386`

Now when I generate a pdf apparently nothing is happening. If I uncomment the lines 140, 144 to 152 I got the initial error again.

**Comment by reinhardt1053** [comment]
I also changed the plugin settings updating the path of wkhtmltopdf

<img width="819" alt="screen shot 2016-09-21 at 16 13 34" src="https://cloud.githubusercontent.com/assets/272154/18714567/7ad16b4c-8016-11e6-9671-c4ec8329297f.png">

The binary wkhtmltopdf-i386 is presented in the specified folder.

**Comment by iusvar** [comment]
Try adding in `snappygrav.php` file after line 30 this:
```
if (!class_exists('Knp\Snappy\Pdf')) {
    $this->grav['debugger']->addMessage( "Knp\Snappy\Pdf doesn't exist! Run 'composer require knplabs/knp-snappy'" );
    return;
} else {
    $this->grav['debugger']->addMessage( "Knp\Snappy\Pdf OK" );
}

$wk_path = ROOT_DIR . $this->config->get('plugins.snappygrav.wk_path');
if (!file_exists($wk_path)) {
    $this->grav['debugger']->addMessage( "wkhtmltopdf doesn't exist! Run 'composer require h4cc/wkhtmltopdf-i386'" );
    return;
} else {
    $this->grav['debugger']->addMessage( "wkhtmltopdf OK" );
}
```
Then make sure that `Admin`, `Configuration`, `System`, `Display errors` & `Debugger` are both Yes:
this allows you to see `Messages` in debug environment...

**Comment by reinhardt1053** [comment]
I added the snippet you suggested but I was getting again the initial error. 

I changed the wkhtmltopdf from a relative path to an absolute path. Now I don't get any error but apparently, nothing happens when I generate the pdf, it redirects to my initial HTML page.

**Comment by reinhardt1053** [comment]
wkhtmltopdf path:
<img width="833" alt="screen shot 2016-09-22 at 10 15 12" src="https://cloud.githubusercontent.com/assets/272154/18741097/9c9f1ae8-80ad-11e6-9011-85a92dee91fe.png">

Debug output:
<img width="1293" alt="screen shot 2016-09-22 at 10 15 30" src="https://cloud.githubusercontent.com/assets/272154/18741101/a21fcaa8-80ad-11e6-81e9-47df36c72040.png">

**Comment by reinhardt1053** [comment]
Output when I try to run wkhtmltopdf by commandline:
<img width="843" alt="screen shot 2016-09-22 at 10 17 10" src="https://cloud.githubusercontent.com/assets/272154/18741147/d1481998-80ad-11e6-839f-bb1d3ec60bc1.png">

**Comment by reinhardt1053** [comment]
<img width="444" alt="screen shot 2016-09-22 at 10 18 55" src="https://cloud.githubusercontent.com/assets/272154/18741183/02a20044-80ae-11e6-8cac-c412a699a9d3.png">

**Comment by iusvar** [comment]
From the plugin folder where you are (`~/www/html/user/plugins/snappygrav`) - when 'binary' is not on the path and not in the current dir - you have to write the whole path, probably: `/home/YOUR_USERNAME/www/html/vendor/h4cc/wkhtmltopdf-i386/bin/wkhtmltopdf-i386`
otherwise - much better - you have to go in bin folder and run `./wkhtmltopdf-i386`

**Comment by reinhardt1053** [comment]
Please check again my latest screenshot, that's exactly what I did, I am using the absolute path
`/home/grav/www/html/vendor/h4cc/wkhtmltopdf-i386/bin/wkhtmltopdf-i386`

My username is grav, I have installed and configured Ubuntu 16.04 server following the official Grav guide: https://learn.getgrav.org/webservers-hosting/vps/digitalocean

**Comment by iusvar** [comment]
You're right, try going directly in the program folder and run `./wkhtmltopdf-i386`

**Comment by reinhardt1053** [comment]
Ok, I had the wrong binary installed, I went on and installed the 64bit version :

`composer require h4cc/wkhtmltopdf-amd64`

And I updated the path on the plugin settings with an absolute path:
`/www/html/vendor/h4cc/wkhtmltopdf-amd64/bin/wkhtmltopdf-amd64`

but it didn't work, debugging the code made it clear that the path must be relative to the grav root. 

Finally I set the path to:

`vendor/h4cc/wkhtmltopdf-amd64/bin/wkhtmltopdf-amd64`

and I got the pdf generated. 

Unrelated question: 
The generated pdf got the images cut horizontally, any way to workaround this issue?

Browser view:
<img width="975" alt="screen shot 2016-09-22 at 12 27 50" src="https://cloud.githubusercontent.com/assets/272154/18744876/0bbd41e0-80c0-11e6-8bc2-aecceaed951b.png">

Pdf generated:

<img width="939" alt="screen shot 2016-09-22 at 12 27 27" src="https://cloud.githubusercontent.com/assets/272154/18744896/1ff4bcce-80c0-11e6-8b44-0b9f01bc018c.png">
